app/Jobs/CreateQueryserviceBatchesJob.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+namespace App\Jobs;
+
+use App\EventPageUpdate;
+use App\QsBatch;
+use App\QsCheckpoint;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Config;
+
+class CreateQueryserviceBatchesJob extends Job
+{
+    private const NAMESPACE_ITEM = 120;
+    private const NAMESPACE_PROPERTY = 122;
+    private const NAMESPACE_LEXEME = 146;
+
+    private int $entityLimit;
+
+    public function __construct()
+    {
+        $this->entityLimit = Config::get('wbstack.qs_batch_entity_limit');
+    }
+
+    public function handle(): void
+    {
+        DB::transaction(function () {
+            $lastCheckpoint = QsCheckpoint::get();
+
+            // Get events after the point that batch was created
+            // TODO maybe filter by NS here?
+            $events = EventPageUpdate::where(
+                'id', '>', $lastCheckpoint
+            )->get();
+
+            $wikiBatchesEntities = [];
+            $latestEventId = $lastCheckpoint;
+            foreach ($events as $event) {
+                if (
+                    $event->namespace == self::NAMESPACE_ITEM ||
+                    $event->namespace == self::NAMESPACE_PROPERTY ||
+                    $event->namespace == self::NAMESPACE_LEXEME
+                ) {
+                    $wikiBatchesEntities[$event->wiki_id][] = $event->title;
+                }
+                if ($event->id > $latestEventId) {
+                    $latestEventId = $event->id;
+                }
+            }
+
+            $notDoneBatches = QsBatch::where([
+                ['done', '=', 0],
+                ['pending_since', '=', null],
+                ['failed', '=', false],
+            ])->get();
+
+            // Insert the newly created batches into the table...
+            foreach ($wikiBatchesEntities as $wikiId => $entityIdsFromEvents) {
+                // If we already have a not done batch for this same wiki, then merge that into a new batch
+                foreach ($notDoneBatches as $qsBatch) {
```

Comment:
maybe we can factor out this loop with a slightly more descriptive name. e.g. `attemptToInsertIntoExistingBatch` or something wittier

Reply:
This went totally out of hand and I refactored basically EVERYTHING :/

I tried to do this in atomic commits though, so if you feel like we should play it safe and drop any of these refactorings (or all of them) please lmk and we shall do so.